man/man1/notmuch-insert.1
.TH NOTMUCH-INSERT 1 2013-12-30 "Notmuch 0.17"
.SH NAME
notmuch-insert \- add a message to the maildir and notmuch database
.SH SYNOPSIS

.B notmuch insert
.RI "[" options "]"
.RI "[ +<" tag> "|\-<" tag "> ... ]"

.SH DESCRIPTION

.B notmuch insert
reads a message from standard input
and delivers it into the maildir directory given by configuration option
.BR database.path ,
then incorporates the message into the notmuch database.
It is an alternative to using a separate tool to deliver
the message then running
.B notmuch new
afterwards.

The new message will be tagged with the tags specified by the
.B new.tags
configuration option, then by operations specified on the command-line:
tags prefixed by '+' are added while
those prefixed by '\-' are removed.

If the new message is a duplicate of an existing message in the database
(it has same Message-ID), it will be added to the maildir folder and
notmuch database, but the tags will not be changed.

Option arguments must appear before any tag operation arguments.
Supported options for
.B insert
include
.RS 4
.TP 4
.BI "--folder=<" folder ">"

Deliver the message to the specified folder,
relative to the top-level directory given by the value of
\fBdatabase.path\fR.
The default is to deliver to the top-level directory.

.RE

.RS 4
.TP 4
.B "--create-folder"

Try to create the folder named by the
.B "--folder"
option, if it does not exist.
Otherwise the folder must already exist for mail
delivery to succeed.

.RE
.SH EXIT STATUS

This command returns exit status 0 if the message was successfully
added to the mail directory, even if the message could not be indexed
and added to the notmuch database.  In the latter case, a warning will
be printed to standard error but the message file will be left on disk.

If the message could not be written to disk then a non-zero exit
status is returned.

.RE
.SH SEE ALSO

\fBnotmuch\fR(1), \fBnotmuch-config\fR(1), \fBnotmuch-count\fR(1),
\fBnotmuch-dump\fR(1), \fBnotmuch-hooks\fR(5), \fBnotmuch-reply\fR(1),
\fBnotmuch-restore\fR(1), \fBnotmuch-search\fR(1),
\fBnotmuch-search-terms\fR(7), \fBnotmuch-show\fR(1),
\fBnotmuch-tag\fR(1)


NEWS
Notmuch 0.17 (2013-12-30)
=========================

Incompatible change in SHA1 computation
---------------------------------------

Previously on big endian architectures like sparc and powerpc the
computation of SHA1 hashes was incorrect. This meant that messages
with overlong or missing message-ids were given different computed
message-ids than on more common little endian architectures like i386
and amd64.  If you use notmuch on a big endian architecture, you are
strongly advised to make a backup of your tags using `notmuch dump`
before this upgrade. You can locate the affected files using something
like:

    notmuch dump | \
      awk '/^notmuch-sha1-[0-9a-f]{40} / \
        {system("notmuch search --exclude=false --output=files id:" $1)}'

Command-Line Interface
----------------------

New options to better support handling duplicate messages

  If more than one message file is associated with a message-id,
  `notmuch search --output=files` will print all of them. A new
  `--duplicate=N` option can be used to specify which duplicate to
  print for each message.

  `notmuch count` now supports `--output=files` option to output the
  number of files associated with matching messages. This may be
  bigger than the number of matching messages due to duplicates
  (i.e. multiple files having the same message-id).

Improved `notmuch new` performance for unchanged folders

  `notmuch new` now skips over unchanged folders more efficiently,
  which can substantially improve the performance of checking for new
  mail in some situations (like NFS-mounted Maildirs).

`notmuch reply --format=text` RFC 2047-encodes headers

  Previously, this used a mix of standard MIME encoding for the reply
  body and UTF-8 for the headers.  Now, the text format reply template
  RFC 2047-encodes the headers, making the output a valid RFC 2822
  message.  The JSON/sexp format is unchanged.

`notmuch compact` command

  The new `compact` command exposes Xapian's compaction
  functionality through a more convenient interface than
  `xapian-compact`. `notmuch compact` will compact the database to a
  temporary location, optionally backup the original database, and
  move the compacted database into place.

Emacs Interface
---------------

`notmuch-tree` (formerly `notmuch-pick`) has been added to mainline

  `notmuch-tree` is a threaded message view for the emacs
  interface. Each message is one line in the results and the thread
  structure is shown using UTF-8 box drawing characters (similar to
  Mutt's threaded view). It comes between search and show in terms of
  amount of output and can be useful for viewing both single threads
  and multiple threads.

  Using `notmuch-tree`

  The main key entries to notmuch tree are

  'z' enter a query to view using notmuch tree (works in hello,
      search, show and tree mode itself)

  'Z' view the current query in tree notmuch tree (works from search
      and show)

  Once in tree mode, keybindings are mostly in line with the rest of
  notmuch and are all viewable with '?' as usual.

  Customising `notmuch-tree`

  `notmuch-tree` has several customisation variables. The most
  significant is the first notmuch-tree-show-out which determines the
  behaviour when selecting a message (with RET) in tree view. By
  default tree view uses a split window showing the single message in
  the bottom pane. However, if this option is set then it views the
  whole thread in the complete window jumping to the selected message
  in the thread. In either case command-prefix selects the other option.

Tagging threads in search is now race-free

  Previously, adding or removing a tag from a thread in a search
  buffer would affect messages that had arrived after the search was
  performed, resulting in, for example, archiving messages that were
  never seen.  Tagging now affects only the messages that were in the
  thread when the search was performed.

`notmuch-hello` refreshes when switching to the buffer

  The hello buffer now refreshes whenever you switch to the buffer,
  regardless of how you get there.  You can disable automatic
  refreshing by customizing `notmuch-hello-auto-refresh`.

Specific mini-buffer prompts for tagging operations

  When entering tags to add or remove, the mini-buffer prompt now
  indicates what operation will be performed (e.g., "Tag thread", "Tag
  message", etc).

Built-in help improvements

  Documentation for many commands has been improved, as displayed by
  `notmuch-help` (usually bound to "?").  The bindings listed by
  `notmuch-help` also now include descriptions of prefixed commands.

Quote replies as they are displayed in show view

  We now render the parts for reply quoting the same way they are
  rendered for show. At this time, the notable change is that replies
  to text/calendar are now pretty instead of raw vcalendar.

Fixed inconsistent use of configured search order

  All ways of interactively invoking search now honor the value of
  `notmuch-search-oldest-first`.

Common keymap for notmuch-wide bindings

  Several key bindings have been moved from mode-specific keymaps to
  the single `notmuch-common-keymap`, which is inherited by each
  notmuch mode.  If you've customized your key bindings, you may want
  to move some of them to the common keymap.

The `notmuch-tag` function now requires a list of tag changes

  For users who have scripted the Emacs interface: the `notmuch-tag`
  API has changed.  Previously, it accepted either a list of tag
  changes or a space-separated string of tag changes.  The latter is
  no longer supported and the function now returns nothing.

Fixed `notmuch-reply` putting reply in primary selection

  On emacs 24 notmuch-reply used to put the cited text into the
  primary selection (which could lead to inadvertently pasting this
  cited text elsewhere). Now the primary-selection is not changed.

Fixed `notmuch-show` invisible part handling


notmuch-insert.c
/* notmuch - Not much of an email program, (just index and search)
 *
 * Copyright © 2013 Peter Wang
 *
 * Based in part on notmuch-deliver
 * Copyright © 2010 Ali Polatel
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/ .
 *
 * Author: Peter Wang <novalazy@gmail.com>
 */

#include "notmuch-client.h"
#include "tag-util.h"

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>

static volatile sig_atomic_t interrupted;

static void
handle_sigint (unused (int sig))
{
    static char msg[] = "Stopping...         \n";

    /* This write is "opportunistic", so it's okay to ignore the
     * result.  It is not required for correctness, and if it does
     * fail or produce a short write, we want to get out of the signal
     * handler as quickly as possible, not retry it. */
    IGNORE_RESULT (write (2, msg, sizeof (msg) - 1));
    interrupted = 1;
}

/* Like gethostname but guarantees that a null-terminated hostname is
 * returned, even if it has to make one up. Invalid characters are
 * substituted such that the hostname can be used within a filename.
 */
static void
safe_gethostname (char *hostname, size_t len)
{
    char *p;

    if (gethostname (hostname, len) == -1) {
	strncpy (hostname, "unknown", len);
    }
    hostname[len - 1] = '\0';

    for (p = hostname; *p != '\0'; p++) {
	if (*p == '/' || *p == ':')
	    *p = '_';
    }
}

/* Call fsync() on a directory path. */
static notmuch_bool_t
sync_dir (const char *dir)
{
    notmuch_bool_t ret;
    int fd;

    fd = open (dir, O_RDONLY);
    if (fd == -1) {
	fprintf (stderr, "Error: open() dir failed: %s\n", strerror (errno));
	return FALSE;
    }
    ret = (fsync (fd) == 0);
    if (! ret) {
	fprintf (stderr, "Error: fsync() dir failed: %s\n", strerror (errno));
    }
    close (fd);
    return ret;
}

/* Check the specified folder name does not contain a directory
 * component ".." to prevent writes outside of the Maildir hierarchy. */
static notmuch_bool_t
check_folder_name (const char *folder)
{
    const char *p = folder;

    for (;;) {
	if ((p[0] == '.') && (p[1] == '.') && (p[2] == '\0' || p[2] == '/'))
	    return FALSE;
	p = strchr (p, '/');
	if (!p)
	    return TRUE;
	p++;
    }
}

/* Make the given directory, succeeding if it already exists. */
static notmuch_bool_t
make_directory (char *path, int mode)
{
    notmuch_bool_t ret;
    char *slash;

    if (mkdir (path, mode) != 0)
	return (errno == EEXIST);

    /* Sync the parent directory for durability. */
    ret = TRUE;
    slash = strrchr (path, '/');
    if (slash) {
	*slash = '\0';
	ret = sync_dir (path);
	*slash = '/';
    }
    return ret;
}

/* Make the given directory including its parent directories as necessary.
 * Return TRUE on success, FALSE on error. */
static notmuch_bool_t
make_directory_and_parents (char *path, int mode)
{
    struct stat st;
    char *start;
    char *end;
    notmuch_bool_t ret;

    /* First check the common case: directory already exists. */
    if (stat (path, &st) == 0)
	return S_ISDIR (st.st_mode) ? TRUE : FALSE;

    for (start = path; *start != '\0'; start = end + 1) {
	/* start points to the first unprocessed character.
	 * Find the next slash from start onwards. */
	end = strchr (start, '/');

	/* If there are no more slashes then all the parent directories
	 * have been made.  Now attempt to make the whole path. */
	if (end == NULL)
	    return make_directory (path, mode);

	/* Make the path up to the next slash, unless the current
	 * directory component is actually empty. */
	if (end > start) {
	    *end = '\0';
	    ret = make_directory (path, mode);
	    *end = '/';
	    if (! ret)
		return FALSE;
	}
    }

    return TRUE;
}

/* Create the given maildir folder, i.e. dir and its subdirectories
 * 'cur', 'new', 'tmp'. */
static notmuch_bool_t
maildir_create_folder (void *ctx, const char *dir)
{
    const int mode = 0700;
    char *subdir;
    char *tail;

    /* Create 'cur' directory, including parent directories. */
    subdir = talloc_asprintf (ctx, "%s/cur", dir);
    if (! subdir) {
	fprintf (stderr, "Out of memory.\n");
	return FALSE;
    }
    if (! make_directory_and_parents (subdir, mode))
	return FALSE;

    tail = subdir + strlen (subdir) - 3;

    /* Create 'new' directory. */
    strcpy (tail, "new");
    if (! make_directory (subdir, mode))
	return FALSE;

    /* Create 'tmp' directory. */
    strcpy (tail, "tmp");
    if (! make_directory (subdir, mode))
	return FALSE;

    talloc_free (subdir);
    return TRUE;
}

/* Open a unique file in the 'tmp' sub-directory of dir.
 * Returns the file descriptor on success, or -1 on failure.
 * On success, file paths for the message in the 'tmp' and 'new'
 * directories are returned via tmppath and newpath,
 * and the path of the 'new' directory itself in newdir. */
static int
maildir_open_tmp_file (void *ctx, const char *dir,
		       char **tmppath, char **newpath, char **newdir)
{
    pid_t pid;
    char hostname[256];
    struct timeval tv;
    char *filename;
    int fd = -1;

    /* We follow the Dovecot file name generation algorithm. */
    pid = getpid ();
    safe_gethostname (hostname, sizeof (hostname));
    do {
	gettimeofday (&tv, NULL);
	filename = talloc_asprintf (ctx, "%ld.M%ldP%d.%s",
				    tv.tv_sec, tv.tv_usec, pid, hostname);
	if (! filename) {
	    fprintf (stderr, "Out of memory\n");
	    return -1;
	}

	*tmppath = talloc_asprintf (ctx, "%s/tmp/%s", dir, filename);
	if (! *tmppath) {
	    fprintf (stderr, "Out of memory\n");
	    return -1;
	}

	fd = open (*tmppath, O_WRONLY | O_CREAT | O_TRUNC | O_EXCL, 0600);
    } while (fd == -1 && errno == EEXIST);

    if (fd == -1) {
	fprintf (stderr, "Error: opening %s: %s\n", *tmppath, strerror (errno));
	return -1;
    }

    *newdir = talloc_asprintf (ctx, "%s/new", dir);
    *newpath = talloc_asprintf (ctx, "%s/new/%s", dir, filename);
    if (! *newdir || ! *newpath) {
	fprintf (stderr, "Out of memory\n");
	close (fd);
	unlink (*tmppath);
	return -1;
    }

    talloc_free (filename);

    return fd;
}

/* Copy the contents of standard input (fdin) into fdout.
 * Returns TRUE if a non-empty file was written successfully.
 * Otherwise, return FALSE. */
static notmuch_bool_t
copy_stdin (int fdin, int fdout)
{
    notmuch_bool_t empty = TRUE;

    while (! interrupted) {
	ssize_t remain;
	char buf[4096];
	char *p;

	remain = read (fdin, buf, sizeof (buf));
	if (remain == 0)
	    break;
	if (remain < 0) {
	    if (errno == EINTR)
		continue;
	    fprintf (stderr, "Error: reading from standard input: %s\n",
		     strerror (errno));
	    return FALSE;
	}

	p = buf;
	do {
	    ssize_t written = write (fdout, p, remain);
	    if (written < 0 && errno == EINTR)
		continue;
	    if (written <= 0) {
		fprintf (stderr, "Error: writing to temporary file: %s",
			 strerror (errno));
		return FALSE;
	    }
	    p += written;
	    remain -= written;
	    empty = FALSE;
	} while (remain > 0);
    }

    return (!interrupted && !empty);
}

/* Add the specified message file to the notmuch database, applying tags.
 * The file is renamed to encode notmuch tags as maildir flags. */
static void
add_file_to_database (notmuch_database_t *notmuch, const char *path,
		      tag_op_list_t *tag_ops, notmuch_bool_t synchronize_flags)
{
    notmuch_message_t *message;
    notmuch_status_t status;

    status = notmuch_database_add_message (notmuch, path, &message);
    switch (status) {
    case NOTMUCH_STATUS_SUCCESS:
    case NOTMUCH_STATUS_DUPLICATE_MESSAGE_ID:
	break;
    default:
    case NOTMUCH_STATUS_FILE_NOT_EMAIL:
    case NOTMUCH_STATUS_READ_ONLY_DATABASE:
    case NOTMUCH_STATUS_XAPIAN_EXCEPTION:
    case NOTMUCH_STATUS_OUT_OF_MEMORY:
    case NOTMUCH_STATUS_FILE_ERROR:
    case NOTMUCH_STATUS_NULL_POINTER:
    case NOTMUCH_STATUS_TAG_TOO_LONG:
    case NOTMUCH_STATUS_UNBALANCED_FREEZE_THAW:
    case NOTMUCH_STATUS_UNBALANCED_ATOMIC:
    case NOTMUCH_STATUS_LAST_STATUS:
	fprintf (stderr, "Error: failed to add `%s' to notmuch database: %s\n",
		 path, notmuch_status_to_string (status));
	return;
    }

    if (status == NOTMUCH_STATUS_DUPLICATE_MESSAGE_ID) {
	/* Don't change tags of an existing message. */
	if (synchronize_flags) {
	    status = notmuch_message_tags_to_maildir_flags (message);
	    if (status != NOTMUCH_STATUS_SUCCESS)
		fprintf (stderr, "Error: failed to sync tags to maildir flags\n");
	}
    } else {
	tag_op_flag_t flags = synchronize_flags ? TAG_FLAG_MAILDIR_SYNC : 0;

	tag_op_list_apply (message, tag_ops, flags);
    }

    notmuch_message_destroy (message);
}

static notmuch_bool_t
insert_message (void *ctx, notmuch_database_t *notmuch, int fdin,
		const char *dir, tag_op_list_t *tag_ops,
		notmuch_bool_t synchronize_flags)
{
    char *tmppath;
    char *newpath;
    char *newdir;
    int fdout;
    char *cleanup_path;

    fdout = maildir_open_tmp_file (ctx, dir, &tmppath, &newpath, &newdir);
    if (fdout < 0)
	return FALSE;

    cleanup_path = tmppath;

    if (! copy_stdin (fdin, fdout))
	goto FAIL;

    if (fsync (fdout) != 0) {
	fprintf (stderr, "Error: fsync failed: %s\n", strerror (errno));
	goto FAIL;
    }

    close (fdout);
    fdout = -1;

    /* Atomically move the new message file from the Maildir 'tmp' directory
     * to the 'new' directory.  We follow the Dovecot recommendation to
     * simply use rename() instead of link() and unlink().
     * See also: http://wiki.dovecot.org/MailboxFormat/Maildir#Mail_delivery
     */
    if (rename (tmppath, newpath) != 0) {
	fprintf (stderr, "Error: rename() failed: %s\n", strerror (errno));
	goto FAIL;
    }

    cleanup_path = newpath;

    if (! sync_dir (newdir))
	goto FAIL;

    /* Even if adding the message to the notmuch database fails,
     * the message is on disk and we consider the delivery completed. */
    add_file_to_database (notmuch, newpath, tag_ops, synchronize_flags);

    return TRUE;

  FAIL:
    if (fdout >= 0)
	close (fdout);
    unlink (cleanup_path);
    return FALSE;
}

int
notmuch_insert_command (notmuch_config_t *config, int argc, char *argv[])
{
    notmuch_database_t *notmuch;
    struct sigaction action;
    const char *db_path;
    const char **new_tags;
    size_t new_tags_length;
    tag_op_list_t *tag_ops;
    char *query_string = NULL;
    const char *folder = NULL;
    notmuch_bool_t create_folder = FALSE;
    notmuch_bool_t synchronize_flags;
    const char *maildir;
    int opt_index;
    unsigned int i;
    notmuch_bool_t ret;

    notmuch_opt_desc_t options[] = {
	{ NOTMUCH_OPT_STRING, &folder, "folder", 0, 0 },
	{ NOTMUCH_OPT_BOOLEAN, &create_folder, "create-folder", 0, 0 },
	{ NOTMUCH_OPT_END, 0, 0, 0, 0 }
    };

    opt_index = parse_arguments (argc, argv, options, 1);
    if (opt_index < 0)
	return EXIT_FAILURE;

    db_path = notmuch_config_get_database_path (config);
    new_tags = notmuch_config_get_new_tags (config, &new_tags_length);
    synchronize_flags = notmuch_config_get_maildir_synchronize_flags (config);

    tag_ops = tag_op_list_create (config);
    if (tag_ops == NULL) {
	fprintf (stderr, "Out of memory.\n");
	return EXIT_FAILURE;
    }
    for (i = 0; i < new_tags_length; i++) {
	if (tag_op_list_append (tag_ops, new_tags[i], FALSE))
	    return EXIT_FAILURE;
    }

    if (parse_tag_command_line (config, argc - opt_index, argv + opt_index,
				&query_string, tag_ops))
	return EXIT_FAILURE;

    if (*query_string != '\0') {
	fprintf (stderr, "Error: unexpected query string: %s\n", query_string);
	return EXIT_FAILURE;
    }

    if (folder == NULL) {
	maildir = db_path;
    } else {
	if (! check_folder_name (folder)) {
	    fprintf (stderr, "Error: bad folder name: %s\n", folder);
	    return EXIT_FAILURE;
	}
	maildir = talloc_asprintf (config, "%s/%s", db_path, folder);
	if (! maildir) {
	    fprintf (stderr, "Out of memory\n");
	    return EXIT_FAILURE;
	}
	if (create_folder && ! maildir_create_folder (config, maildir)) {
	    fprintf (stderr, "Error: creating maildir %s: %s\n",
		     maildir, strerror (errno));
	    return EXIT_FAILURE;
	}
    }

    /* Setup our handler for SIGINT. We do not set SA_RESTART so that copying
     * from standard input may be interrupted. */
    memset (&action, 0, sizeof (struct sigaction));
    action.sa_handler = handle_sigint;
    sigemptyset (&action.sa_mask);
    action.sa_flags = 0;
    sigaction (SIGINT, &action, NULL);

    if (notmuch_database_open (notmuch_config_get_database_path (config),
			       NOTMUCH_DATABASE_MODE_READ_WRITE, &notmuch))
	return EXIT_FAILURE;

    ret = insert_message (config, notmuch, STDIN_FILENO, maildir, tag_ops,
			  synchronize_flags);

    notmuch_database_destroy (notmuch);

    return ret ? EXIT_SUCCESS : EXIT_FAILURE;
}
